### PR TITLE
`DateTime`/`DateTime64` support in `toDaysSinceYearZero()`

### LINKED ISSUE #54796: MySQL compatibility: Illegal type of argument #1 'date' of function toDaysSinceYearZero
Required for Tableau Online. 

Sample rendered query:

```sql
SELECT SUM(`cell_towers`.`mcc`)   AS `sum_mcc_ok`,
       ADDDATE(FROM_DAYS(TO_DAYS(`cell_towers`.`updated`) - (DAYOFWEEK(`cell_towers`.`updated`) - 1)),
               INTERVAL 0 SECOND) AS `twk_updated_ok`
FROM `cell_towers`
GROUP BY 2;
```

fails with

```
Code: 43. DB::Exception: Illegal type of argument #1 'date' of function toDaysSinceYearZero, expected Date or Date32, got DateTime: In scope SELECT SUM(cell_towers.mcc) AS sum_mcc_ok, ADDDATE(FROM_DAYS(TO_DAYS(cell_towers.updated) - (DAYOFWEEK(cell_towers.updated) - 1)), toIntervalSecond(0)) AS twk_updated_ok FROM cell_towers GROUP BY 2. (ILLEGAL_TYPE_OF_ARGUMENT)
```

Reduced example:

```sql
SELECT TO_DAYS(`cell_towers`.`updated`) FROM cell_towers LIMIT 1;
```

**How to reproduce**

* Which ClickHouse server version to use: latest master
* Which interface to use, if matters: MySQL
* Sample data for all these tables: cell_towers sample dataset

**Expected behavior**

TO_DAYS accepts DateTime/DateTime64 to be more in line with the MySQL standard.

### PR DESCRIPTION
**Pull request by slvrtrn**
This PR adds support for `DateTime` / `DateTime64` arguments to SQL function `toDaysSinceYearZero()` (aka MySQL `TO_DAYS`). It additionally fixes the calculation itself which was based on a wrong constant (that's not in the changelog as the function is new and was not included in a release yet).

Resolves #54796.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Function `toDaysSinceYearZero()` now supports arguments of type `DateTime` and `DateTime64`.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 44ee98ed8f612aa463e107de28fd8431892dbe33 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54856/44ee98ed8f612aa463e107de28fd8431892dbe33/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Functions/toDaysSinceYearZero.cpp [hunk 1]
```diff
@@ -58,38 +57,41 @@ class FunctionToDaysSinceYearZero : public IFunction
...
-            return DateTimeTransformImpl<DataTypeDate, ResultType, ToDaysSinceYearZeroImpl>::execute(arguments, result_type, input_rows_count);
+            return DateTimeTransformImpl<DataTypeDate, ResultType, ToDaysSinceYearZeroImpl>::execute(
+                arguments, result_type, input_rows_count);
         else if (which.isDate32())
-            return DateTimeTransformImpl<DataTypeDate32, ResultType, ToDaysSinceYearZeroImpl>::execute(arguments, result_type, input_rows_count);
+            return DateTimeTransformImpl<DataTypeDate32, ResultType, ToDaysSinceYearZeroImpl>::execute(
+                arguments, result_type, input_rows_count);
         else if (which.isDateTime())
-            return DateTimeTransformImpl<DataTypeDateTime, ResultType, ToDaysSinceYearZeroImpl>::execute(arguments, result_type, input_rows_count);
+            return DateTimeTransformImpl<DataTypeDateTime, ResultType, ToDaysSinceYearZeroImpl>::execute(
```
Comment (rschu1ze) [inline src/Functions/toDaysSinceYearZero.cpp:66]: Needs tests: tests/queries/0_stateless/02874_toDaysSinceYearZero.sql

File: src/Functions/toDaysSinceYearZero.cpp [hunk 2]
```diff
@@ -58,38 +57,41 @@ class FunctionToDaysSinceYearZero : public IFunction
...
         if (which.isDate())
-            return DateTimeTransformImpl<DataTypeDate, ResultType, ToDaysSinceYearZeroImpl>::execute(arguments, result_type, input_rows_count);
+            return DateTimeTransformImpl<DataTypeDate, ResultType, ToDaysSinceYearZeroImpl>::execute(
+                arguments, result_type, input_rows_count);
         else if (which.isDate32())
-            return DateTimeTransformImpl<DataTypeDate32, ResultType, ToDaysSinceYearZeroImpl>::execute(arguments, result_type, input_rows_count);
+            return DateTimeTransformImpl<DataTypeDate32, ResultType, ToDaysSinceYearZeroImpl>::execute(
+                arguments, result_type, input_rows_count);
         else if (which.isDateTime())
-            return DateTimeTransformImpl<DataTypeDateTime, ResultType, ToDaysSinceYearZeroImpl>::execute(arguments, result_type, input_rows_count);
```
Comment (rschu1ze) [inline src/Functions/toDaysSinceYearZero.cpp]: I don't get it. This is not the code state in branch `master`. What code state is this based on?

Comment (rschu1ze) [reply src/Functions/toDaysSinceYearZero.cpp]: Oh, I looked only at a single commit.

File: src/Functions/toDaysSinceYearZero.cpp (hunk 1 above)
Comment (slvrtrn) [reply src/Functions/toDaysSinceYearZero.cpp:66]: yep, it is there already: :)
https://github.com/ClickHouse/ClickHouse/blob/4d3b127a1ff370fd5cee3dccdc2456163ad816ed/tests/queries/0_stateless/02874_toDaysSinceYearZero.sql

**Comment by qoega** [comment]
@rschu1ze @slvrtrn Let's add `fromDaysSinceYearZero` as well? It looks logical even if there is no such function in mysql

**Comment by rschu1ze** [comment]
@qoega --> https://github.com/ClickHouse/ClickHouse/issues/55308

**Comment by bhavuk2002** [comment]
@qoega there exist a functionality `FROM_DAYS()` in MYSQL which convert number of days to date, [check here](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_from-days).

**Comment by qoega** [comment]
It is good as an alias.

**Comment by bhavuk2002** [comment]
it is similar to what we want to implement here, right?

**Comment by rschu1ze** [comment]
Yes it is. Basically the inverse of `toDaysSinceYearZero()` aka. `TO_DAYS()`.

**Comment by bhavuk2002** [comment]
@rschu1ze 

Its kind of seem to much stuff for me to parse, can you tell me how to approach. I can make a logic to convert number of days to year but your guidance will help.

**Pull request merged**